components/flow/flow-service/src/main/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImpl.kt

```diff
@@ -154,66 +151,59 @@ class ExternalEventManagerImpl(
-                            externalEventState.eventToSend.timestamp
-                }
-                if (externalEventState.status.type == ExternalEventStateType.OK) {
-                    externalEventState.status.exception =
-                        ExceptionEnvelope(
-                            "NoResponse",
-                            "Received no response for external event request, ensure all workers are running"
-                        )
-                    externalEventState.status.type = ExternalEventStateType.RETRY
-                    externalEventState.retries = externalEventState.retries.inc()
-                }
-                getAndUpdateEventToSend(externalEventState, instant, config)
+            ExternalEventStateType.RETRY -> {
+                checkRetry(externalEventState, instant, retryWindow)
+                generateRecord(externalEventState, instant)
+            }
+            else -> {
+                null
             }
-
-            else -> externalEventState to null
         }
+        return externalEventState to record
     }
 
-    private fun hasNotSentOriginalEvent(externalEventState: ExternalEventState): Boolean {
-        return externalEventState.sendTimestamp == null
-    }
-
-    private fun canRetryEvent(externalEventState: ExternalEventState, instant: Instant): Boolean {
-        return if (externalEventState.status.type !in setOf(
-                ExternalEventStateType.PLATFORM_ERROR,
-                ExternalEventStateType.FATAL_ERROR
-            )
-        ) {
-            val sendTimestamp = externalEventState.sendTimestamp.truncatedTo(ChronoUnit.MILLIS).toEpochMilli()
-            val currentTimestamp = instant.truncatedTo(ChronoUnit.MILLIS).toEpochMilli()
-            sendTimestamp < currentTimestamp
-        } else {
-            false
+    private fun checkRetry(externalEventState: ExternalEventState, instant: Instant, retryWindow: Duration) {
+        when {
+            (externalEventState.sendTimestamp + retryWindow) >= instant -> {
+                // Do nothing. This check ensures that subsequent branches are checking the case where the external
+                // event is outside the retry window.
+            }
+            externalEventState.retries == 0 -> {
+                // Use the retries field to indicate how many times the event has been retried outside the window.
+                // Retrying once outside the window is required in case the flow engine receives the event to trigger
+                // the retry late. This guarantees an external event will be tried at least twice. After that though,
+                // retrying further is unlikely to clear the problem.
+                externalEventState.retries++
+            }
+            else -> {
+                throw FlowFatalException("External event with request ID ${externalEventState.requestId} exceeded " +
+                        "the retry window.")
+            }
         }
     }
```

Comment:
I like the use of the `when` block here, but gating the logic with a 'pitfall' first case feels a little unexpected. Might be more clear to just use an early return, e.g.
```suggestion
    private fun checkRetry(externalEventState: ExternalEventState, instant: Instant, retryWindow: Duration) {
        if ((externalEventState.sendTimestamp + retryWindow) >= instant) return
        
        if (0 == externalEventState.retries) {
            // Retrying once outside the window is required in case the flow engine receives the event to trigger
            // the retry late. This guarantees an external event will be tried at least twice. After that though,
            // retrying further is unlikely to clear the problem.
            externalEventState.retries++
        } else {
            throw FlowFatalException("External event with request ID ${externalEventState.requestId} exceeded " +
                        "the retry window.")
        }
    }
```

Reply:
It's an interesting one. It's actually not an early return at all, but an explicit return vs an implicit one. Personally I find it easier to reason that all branches have an implicit return in, but I agree having a branch with no code in is a little surprising (hence the comment).

I'm going to leave this for now I think, but thanks for the comment, it's an interesting question!